fineract-provider/src/main/java/org/apache/fineract/infrastructure/dataqueries/service/ReadReportingServiceImpl.java

```diff
@@ -151,21 +166,45 @@ public GenericResultsetData retrieveGenericResultset(final String name, final St
             final boolean isSelfServiceUserReport) {
 
         final long startTime = System.currentTimeMillis();
-        if (log.isDebugEnabled()) {
-            log.debug("STARTING REPORT: {}   Type: {}", LogParameterEscapeUtil.escapeLogParameter(name),
-                    LogParameterEscapeUtil.escapeLogParameter(type));
+        LOG.info("STARTING REPORT: {}   Type: {}", name, type);
+        StringBuilder sqlStringBuilder = new StringBuilder(200);
+        sqlStringBuilder.append(getSQLtoRun(name, type, queryParams, isSelfServiceUserReport));
+        final GenericResultsetData result;
+        boolean isPaginationAllowed = Boolean.parseBoolean(queryParams.get(ReportingConstants.IS_PAGINATION_ALLOWED));
+        Page<GenericResultsetData> reportData = this.paginationHelper.fetchPage(this.jdbcTemplate, sqlStringBuilder.toString(), null,
+                new ReportMapper(sqlStringBuilder));
+        if (isPaginationAllowed) {
+            sqlStringBuilder = retrieveGenericResultsetWithPagination(sqlStringBuilder, queryParams);
         }
+        result = this.genericDataService.fillGenericResultSet(sqlStringBuilder.toString());
+        int pageSize = this.configurationDomainService.reportsPaginationNumberOfItemsPerPage();
 
-        final String sql = getSQLtoRun(name, type, queryParams, isSelfServiceUserReport);
+        final long elapsed = System.currentTimeMillis() - startTime;
+        LOG.info("FINISHING Report/Request Name: {} - {}     Elapsed Time: {}", name, type, elapsed);
+        return GenericResultsetData.setTotalItemsAndRecordsPerPage(result, reportData.getTotalFilteredRecords(), pageSize);
+    }
 
-        final GenericResultsetData result = this.genericDataService.fillGenericResultSet(sql);
+    public StringBuilder retrieveGenericResultsetWithPagination(final StringBuilder sqlStringBuilder,
```

Comment:
This method is lying. It's not retrieving a generic resultset with pagination. It's constructing the SQL for it.